### PR TITLE
[SYCL] Fixed device selection for SYCL_DEVICE_FILTER

### PR DESCRIPTION
**Pull request by nawrinsu**
This patch fixes the device selection when SYCL_DEVICE_FILTER=
level_zero:cpu is used. As such a device does not exists,
currently the default_selector selects the level_zero GPU
device. This patch fixes the selection by chosing the only
available device HOST in this case.

Signed-off-by: Nawrin Sultana <nawrin.sultana@intel.com>

### CODE REVIEW

File: sycl/test/on-device/select_device_host.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+      device d = ds.select_device();
+      string name = d.get_platform().get_info<info::platform::name>();
+      assert(name.find("host") != string::npos);
+      std::cout << "Host Device is selected: " << boolalpha << d.is_host() << std::endl;
+    }
+    {
+      gpu_selector gs;
+      try {
+        device d = gs.select_device();
+        cerr << "CPU device is found in error: " << d.is_cpu() << std::endl;
```
Comment (s-kanaev) [inline sycl/test/on-device/select_device_host.cpp:29]:
```suggestion
        cerr << "GPU device is found in error: " << d.is_gpu() << std::endl;
```

Is this what was intended?

**Review by s-kanaev — COMMENTED** [review]
Looks good overall

**Comment by s-kanaev** [comment]
@bso-intel , could you, please, review?

File: sycl/test/on-device/select_device_host.cpp (hunk 1 above)
Comment (nawrinsu) [reply sycl/test/on-device/select_device_host.cpp]: It was a typo. I fixed it.

**Comment by bader** [comment]
> @bso-intel , could you, please, review?

@s-kanaev, please, use GitHub UI to request the review by adding @bso-intel to reviewers.

**Review by bso-intel — COMMENTED** [review]
It seems there was some misunderstanding about the SYCL_DEVICE_FILTER.
It has two purposes.
1. Set the desired device
2. Limit loading of supported plugins into the SYCL RT.
Setting SYCL_DEVICE_FILTER=level_zero:cpu means the user wants to use the level_zero:cpu device so the SYCL RT has to load the level_zero backend which can support only GPU device at this moment, but it can support more device types in the future.
default-selector is heuristic based and it does not necessarily choose the device designated by SYCL_DEVICE_FILTER.
The filter only plays as the **limiting** filter out of all available options.
Since we are only allowed to load the level_zero backend, the highest heuristic point is the GPU device from all available devices (GPU, HOST).
So, the observed behavior of the default selector is correct. (GPU).
I was planning to add some warning message to the user if default selector cannot find any matched device with SYCL_DEVICE_FILTER.

**Comment by bso-intel** [comment]
> Limits the SYCL RT to use only a subset of the system's devices. 

Sorry, I think we should clarify this point better in the documentation.
As I mentioned above, this sentence means SYCL_DEVICE_FILTER only limits the available devices into a subset. 
The available devices are determined by user's backend choice, i.e., "level_zero". Level Zero supports only GPU devices today. HOST is always available. So, when the user sets SYCL_DEVICE_FILTER=level_zero:*, SYCL RT only has two choices, GPU and HOST.
And depending on the user's device_selector choice, the returned device varies. (details are in the next section below)

> Setting this environment variable affects all of the device query functions and all of the device selectors.

Please note that there are different device selectors and their basic behaviors are the same as before. The default_selector always behaves as the heuristic based device selector. Among all the available devices, the default_selector will try to match the SYCL_DEVICE_FILTER. CPU device is out of the option since the user designated Level Zero backend. Between GPU and HOST, GPU has the higher heuristic point, so that's why it is chosen. Please note that this behavior is not different from previous definition of default_selector.
There are other non-general device_selectors (such as cpu_selector, gpu_selector). And when the user uses cpu_selector with SYCL_DEVICE_FILTER=level_zero:cpu, it will throw an exception. Again, this has been the behavior we have implemented before, and SYCL_DEVICE_FILTER only act as a limiting filter. For example, when the user sets SYCL_DEVIE_FILTER=level_zero and use gpu_selector, it returns a level_zero gpu device.

I am planning to emit a warning message in the heuristic-base default_selector when SYCL RT cannot find a designated device_type is found. For example, "WARNING: SYCL_DEVICE_FILTER was set to level_zero:cpu, but we could not find the exact matched device, so a GPU device, the best match, is chosen." I think something like this can help users avoid confusion about this behavior.

**Comment by steffenlarsen** [comment]
I agree with Vlad. It wouldn't be a filter, as the name suggests, if it's just setting a preference, so to not betray the naming it should only allow selecting anything that makes it through the filter.

Maybe an additional option could be something along the lines of `SYCL_DEVICE_PREFERNCE`. For example, say I only wanted to use OpenCL and preferred to use GPU, but would be fine with CPU, then I'd set `SYCL_DEVICE_FILTER=opencl` and `SYCL_DEVICE_PREFERNCE=gpu`. That is outside the scope of this MR though.

**Comment by bso-intel** [comment]
> I agree with @romanovvlad . It wouldn't be a filter, as the name suggests, if it's just setting a preference, so to not betray the naming it should only allow selecting anything that makes it through the filter.
> 
> Maybe an additional option could be something along the lines of `SYCL_DEVICE_PREFERNCE`. For example, say I only wanted to use OpenCL and preferred to use GPU, but would be fine with CPU, then I'd set `SYCL_DEVICE_FILTER=opencl` and `SYCL_DEVICE_PREFERNCE=gpu`. That is outside the scope of this MR though.

Your desired behavior "I only wanted to use OpenCL and preferred to use GPU, but would be fine with CPU" can be achieved by setting SYCL_DEVIDE_FILTER=opencl. The default_selector is heuristic based, and it will try to find all devices available from the opencl backend, out of which the GPU device (if found) will be chosen because that has the highest heuristic point. Otherwise, a CPU device will be selected. You cannot achieve this heuristic behavior if you use a gpu_selector or a cpu_selector.

**Comment by steffenlarsen** [comment]
> Your desired behavior "I only wanted to use OpenCL and preferred to use GPU, but would be fine with CPU" can be achieved by setting SYCL_DEVIDE_FILTER=opencl. The default_selector is heuristic based, and it will try to find all devices available from the opencl backend, out of which the GPU device (if found) will be chosen because that has the highest heuristic point. Otherwise, a CPU device will be selected. You cannot achieve this heuristic behavior if you use a gpu_selector or a cpu_selector.

That was not my desired behavior, but rather just an example for an additional option to give further control, then giving an example of how that could be used in the same way as the `SYCL_DEVICE_FILTER` you're describing.

In fact, I can imagine cases where I'd rather it fail if there is no GPU device present, thus I'm advocating for the `SYCL_DEVICE_FILTER` to be a strict filter as its name suggests. It also seems more consistent to me than having it filter backends, but only set preferences for device types.

If someone wants to be able to set preferences in the future, then (in another MR) someone might want to add an option for setting preference, using an option with a name that conveys it.

**Comment by kevin-harms** [comment]
This is from the documentation of SYCL_DEVICE_FILTER - "Assuming a filter has all three elements of the triple, it selects only those devices that come from the given backend, have the specified device type, AND have the given device index."

This seems to indicate that level_zero:cpu should return an empty set because level_zero does not have the "specified device type". Thus the returned value should only be HOST as that is indicated as a default that is returned with all sets.

**Comment by bso-intel** [comment]
@steffenlarsen @kevin-harms 
What you have described applies to non-generic device selectors.
For example, you use gpu_selector while you ask a cpu device type in the filter, then the gpu_selector will throw an exception as you expected.
default_selector is not designed to be bound to a specific device type.
It is designed to be flexible based on the heuristic points of available devices. 
Depending on the system, it can return a GPU device, a CPU device, or a HOST device. And that's what default_selector is designed for.

**Comment by steffenlarsen** [comment]
> @steffenlarsen @kevin-harms
> What you have described applies to non-generic device selectors.
> For example, you use gpu_selector while you ask a cpu device type in the filter, then the gpu_selector will throw an exception as you expected.
> default_selector is not designed to be bound to a specific device type.
> It is designed to be flexible based on the heuristic points of available devices.
> Depending on the system, it can return a GPU device, a CPU device, or a HOST device. And that's what default_selector is designed for.

That is a good point, but say I was given a binary that used the default selector, but for my case I didn't want it to use anything other than my OpenCL GPU. I would expect the `SYCL_DEVICE_FILTER` to exclude everything other than what is in the filter. As @kevin-harms points out, the documentation tells me the same.

Arguably the above scenario is slightly elaborate, but what I'm trying to say is that this is named and documented as a filter, so I advocate for it being just that.

**Comment by bso-intel** [comment]
@steffenlarsen 
Yes, we named this env var as a filter, and it is exactly meant to act as a filter as current implementation does.

_This env var is not meant to deflect the meaning of default_selector._

In your usage scenario, you can set SYCL_DEVICE_FILTER=opencl:gpu, an OpenCL GPU will be chosen by the default_selector because that is the highest heuristic point.

It seems people got mixed up between SYCL_DEVICE_FILTER (acting as a limiting filter) and the definition of default_selector.
Let me repeat.
default_selector is meant to be heuristic based, and it can choose whichever device that has the highest point among all available devices.
The SYCL_DEVICE_FILTER is just literally used as a **filter** to limit these available devices that are supported by designated backend (if any).

I heard clearly the potential problems. I will have some follow-up action items to alleviate the issues.
1. Clarify the impact to various device selectors in the documentation with several usage examples.
2.  Emit a warning message to the user when default_selector selects a novice other than SYCL_DEVICE_FILTER designated.

**Comment by kevin-harms** [comment]
I think everyone agrees that the system behaves as you described, but the behavior isn't expected based on the documentation. Everyone seems to understand and that the filter is: { backend ∩ device_type ∩ index } however your description is { backend ∪ device_type ∪ index }. I think just adding this mathematical definition in the documentation will be clear to everyone.

**Comment by bso-intel** [comment]
@kevin-harms 
The filter actually behaves as intersections of those entries.
It seems I did not clarify enough on how SYCL_DEVICE_FILTER affects the default_selector.
So, let me try again.
The default selector receives all available devices from the specified backend.
When the user sets "level_zero", currently L0 backend provides only GPU device.
In addition, we automatically add a HOST device per SYCL spec.
Now, the default selector checks if SYCL_DEVICE_FILTER was set, and if so, it will play as a limiting filter. (by intersection of filter entries).
The default selector cannot find the CPU device (which could be the highest heuristic point if L0 backend supported it).
So, as default_selector is designed, it will look for the next highest heuristic point, and that is the L0 GPU device.

I understand this could be a little hazy, so I will clarify in the documentation how it affects the various device selectors.
In summary, SYCL_DEVICE_FILTER will act as a limiting factor to satisfy the user's designation.
If a match is found, it will be the returned device as the highest heuristic point.
Otherwise, the existing behaviors of various selectors as described in the spec will continue to hold.

**Comment by TApplencourt** [comment]
Thanks a lot, @bso-intel  for explaining the behavior implemented! 

Should we open a issue to get a  new `$ENV` where the filtering is done before the heuristic of the selector kick in? This will be used to forcing the code to run on a specific triplet and bailout if not available (regarding the selector used). This seems to be a use case that people want (or at least expect).

For example, I expect people to use this variable to map the SYCL device to MPI rank ( each process will have his own GPU). They will not likely use a `gpu_selector` for portability reason (maybe this code run on another system on the CPU). 

 If they put a too large index in the GPU triplet by mistake (for example `OpenCL:GPU:10`)  I'm sure that they will like the code to fail and not to run on a default GPU (as it expected, if my understanding of the SYCL_DEVICE_FILTER is correct). 

I hope this uses case makes sense,
Regards

**Comment by bader** [comment]
> I'll modify the PR to emit a warning message instead of filtering out the backend.

@nawrinsu. do you have any updates on that?

**Comment by bso-intel** [comment]
@nawrinsu , I suggest you close this PR as I have fixed this issue as a part of https://github.com/intel/llvm/pull/3611.

**Comment by bader** [comment]
Okay. I'm closing this pull request. @nawrinsu, feel free to re-open if you think the issue is still not fixed.
